src/compiler/checker.ts

```diff
@@ -22624,13 +22624,13 @@ namespace ts {
                 if (propName === undefined) {
                     return type;
                 }
-                const includesUndefined = strictNullChecks && maybeTypeOfKind(type, TypeFlags.Undefined);
-                const removeOptional = includesUndefined && isOptionalChain(access);
-                let propType = getTypeOfPropertyOfType(removeOptional ? getTypeWithFacts(type, TypeFacts.NEUndefined) : type, propName);
+                const includesNullable = strictNullChecks && maybeTypeOfKind(type, TypeFlags.Nullable);
+                const removeNullable = includesNullable && isOptionalChain(access);
+                let propType = getTypeOfPropertyOfType(removeNullable ? getTypeWithFacts(type, TypeFacts.NEUndefinedOrNull) : type, propName);
                 if (!propType) {
                     return type;
                 }
-                propType = removeOptional ? getOptionalType(propType) : propType;
+                propType = removeNullable ? getOptionalType(propType) : propType;

```

Comment:
Optional chaining leverages an `optionalType` marker type to indicate the type of an expression came from a `?.`. The `optionalType` marker is essentially `undefined`, but its presence has meaning to completions as well as within the checker itself. The code here (and from #42450), removes `undefined` (regardless of source), then mixes back in `undefinedType`. In functions like `checkPropertyAccessChain`, I use the function `propagateOptionalTypeMarker` to restore `optionalType` to the result if it was found in the source type to carry that marker forward.

I'm not certain yet whether this (or #42450) will cause problems as a result of not restoring the `optionalType` marker, as I need to consider if there is a test case that might trigger unwanted behavior.